### PR TITLE
Use a less complicated implementation for the MAKE_ENUM of CONTROLLER.

### PR DESCRIPTION
**Pull request by Pentarctagon**

### CODE REVIEW

**Review by Vultraz — APPROVED** [review]

**Comment by Pentarctagon** [comment]
Overall this seems to be a pretty straight forward change. The one thing that caused a problem initially was that assigning a MAKE_ENUM value to a config attribute sets the string value of it, whereas if it's a regular enum it of course sets it to its numeric value.

The benefits being this new approach is far simpler and easier to understand than trying to delve into how MAKE_ENUM works, as well as having a non-throwing way to check if a string value is a valid member of the associated enum.

Also, barring any issues with this approach, I'd probably (over the course of 1.17) convert other uses of MAKE_ENUM to this approach with the goal of eventually removing MAKE_ENUM entirely.

**Comment by stevecotton** [comment]
Is there a bug that requires this to be merged soon, or can it wait until later in the dev cycle, so that it's less likely that there will be backports to 1.16.x after this merges?

> Also, barring any issues with this approach, I'd probably (over the course of 1.17) convert other uses of MAKE_ENUM to this approach with the goal of eventually removing MAKE_ENUM entirely.

Although it's easier to understand, this gives it its own handcoded class instead of sharing the common code of MAKE_ENUM. I'd prefer to have a template or even a macro, so that other enums that are converted to the new system will share common code instead of being cut&pastes.

File: src/gui/dialogs/game_stats.cpp
Comment (stevecotton) [inline src/gui/dialogs/game_stats.cpp:71]: The game crashes when opening the stats dialog. Note the line preceding it, vs the changed indexes in the new enum.

File: src/gui/dialogs/multiplayer/mp_join_game.cpp
Comment (stevecotton) [inline src/gui/dialogs/multiplayer/mp_join_game.cpp:230]: Both the old and the new code feel wrong - the MAKE_ENUM framework supports converting to and from strings, so why isn't this using it?

File: src/menu_events.cpp
Comment (stevecotton) [inline src/menu_events.cpp:1493]: The original code seems to be making local aliases to avoid the long enum_to_string code. Now that the right-hand-side is just side_controller::human, I don't think the alias is needed.

File: src/enum_maps/side_controller.hpp
Comment (stevecotton) [inline src/enum_maps/side_controller.hpp:28]: Having an `enum` rather than an `enum class`, combined with reusing the same names (except for a different case) for `char*` consts sounds likely to cause bugs where the wrong case is used, but the compiler silently converts `char *` to int.

File: src/enum_maps/side_controller.hpp
Comment (stevecotton) [inline src/enum_maps/side_controller.hpp:48]: I feel the MAKE_ENUM macro's technique of putting each pair on the same line is better than splitting them across separate definitions for enum, strings and mapping array.

File: src/gui/dialogs/multiplayer/mp_join_game.cpp
Comment (Pentarctagon) [reply src/gui/dialogs/multiplayer/mp_join_game.cpp:230]: I don't know why a variety of places aren't using the CONTROLLER MAKE_ENUM in places like these. Also note that the wesnothd and client MAKE_ENUMs aren't even common code - they're copy/pastes of the same few lines into two separate headers.

File: src/enum_maps/side_controller.hpp
Comment (Pentarctagon) [reply src/enum_maps/side_controller.hpp:48]: How would you prefer this be done instead then?

**Comment by Pentarctagon** [comment]
> Is there a bug that requires this to be merged soon, or can it wait until later in the dev cycle, so that it's less likely that there will be backports to 1.16.x after this merges?

There is not. However I would prefer this to be merged sooner than later so that it doesn't accumulate merge conflicts over the coming months.

> Although it's easier to understand, this gives it its own handcoded class instead of sharing the common code of MAKE_ENUM. I'd prefer to have a template or even a macro, so that other enums that are converted to the new system will share common code instead of being cut&pastes.

The problem with MAKE_ENUM is that having common code is not especially useful if the common code is incomprehensible without digging through a bunch of Boost macros+documentation.

**Comment by Pentarctagon** [comment]
> I'm missing something here, because users of MAKE_ENUM shouldn't need to look at the innards of the macro. If new functionality could be added to the existing MAKE_ENUM, would that satisfy your needs?

I could of course simply ignore it and move on, since technically MAKE_ENUM allows supporting the usecases it's needed for, but I also think it's a severely over complicated way to implement the functionality of "let's associate an enum and a string". I don't personally think it's worth anyone's time to add more functionality to or on top of MAKE_ENUM though.

> There's already an overload string_to_enum(const string&, NAME fallback) and a bool parse(templated string) in MAKE_ENUM that don't throw.

`bool parse(templated string)` I ignored since it requires an instance of the MAKE_ENUM to exist already, which it modifies if the string is a valid value. The overloaded `string_to_enum(const string&, NAME fallback)` I missed though.

File: src/gui/dialogs/game_stats.cpp
Comment (Pentarctagon) [reply src/gui/dialogs/game_stats.cpp]: Fixed.

File: src/menu_events.cpp
Comment (Pentarctagon) [reply src/menu_events.cpp]: Fixed.

File: src/enum_maps/side_controller.hpp
Comment (Pentarctagon) [reply src/enum_maps/side_controller.hpp:28]: Fixed.

File: src/enum_maps/side_controller.hpp
Comment (Pentarctagon) [reply src/enum_maps/side_controller.hpp:37]: Yeah, that's left over from when I was initially thinking of using `map`s rather than the approach I went with. Not sure what a better name would be, maybe `string_enums`?

File: src/enum_maps/side_controller.hpp
Comment (Pentarctagon) [reply src/enum_maps/side_controller.hpp]: Updated to `to_string()`.

For the interface, not generally, though having to do:
```
team::CONTROLLER::enum_to_string(team::CONTROLLER::AI);
```
in menu_events.cpp for example to get the string felt rather cumbersome.

File: src/scripting/lua_team.cpp
Comment (stevecotton) [inline src/scripting/lua_team.cpp:83]: A question to possibly defer answering until a different PR: could this use `return_string_attrib` and omit the `.c_str()`? This is one of only two callers of `return_cstring_attrib`.

Comment (Pentarctagon) [reply src/scripting/lua_team.cpp:83]: I'd guess probably, but I don't know for sure. Maybe @CelticMinstrel would know though.

Comment (CelticMinstrel) [reply src/scripting/lua_team.cpp:83]: Yeah, I think that should work.

**Comment by CelticMinstrel** [comment]
I don't quite get the point of this when `MAKE_ENUM` works perfectly well, but whatever.

Also, I don't agree with the decision to add an `INVALID` value. In my opinion, enums should only have the values that are valid. If in one place (such as `get_enum`) you need to signal that the result value is invalid, then I would just return `nullopt`.

**Comment by Vultraz** [comment]
> then I would just return `nullopt`.

I second this. Now that we have `std::optional` a `nullopt` should be returned anywhere an invalid value is allowed instead of a bool or some other invalid value flag.

File: src/gui/dialogs/game_stats.cpp
Comment (Pentarctagon) [reply src/gui/dialogs/game_stats.cpp]: Unintentionally; fixed.

File: src/scripting/lua_team.cpp
Comment (Pentarctagon) [reply src/scripting/lua_team.cpp]: Alright, changed then.

**Comment by Pentarctagon** [comment]
Updated to use std::optional instead of an INVALID enum value.

**Comment by Vultraz** [comment]
Also, one last thing, you could probably use `std::string_view` instead of `const char*` (it has constexpr ctors).

**Comment by Pentarctagon** [comment]
Does config handle string_view automatically, since a string_view can't be implicitly converted to string like const char* can?

This would also end up being rather cumbersome on the server-side with simple_wml since string_view doesn't guarantee its data is null terminated: https://en.cppreference.com/w/cpp/string/basic_string_view/data
> Unlike std::basic_string::data() and string literals, data() may return a pointer to a buffer that is not null-terminated. Therefore it is typically a mistake to pass data() to a routine that takes just a const CharT* and expects a null-terminated string.

While simple_wml only accepts attributes with the types of int, const char*, and string_span.

**Comment by CelticMinstrel** [comment]
I'm also curious what "reserved" is for?

**Comment by Vultraz** [comment]
> Does config handle string_view automatically, since a string_view can't be implicitly converted to string like const char* can?

`std::string` has a ctor that takes a `string_view`.

**Comment by Pentarctagon** [comment]
IIRC "Reserved" is for when someone is hosting a reloaded game that had other human players in it.

**Comment by Vultraz** [comment]
(That is, you should be able to just do `return std::string(return values_[static_cast<int>(key)]);` if it doesn't do that implicitly.)

**Comment by Vultraz** [comment]
Why is that a problem?

**Comment by Pentarctagon** [comment]
Everywhere that would assign it to a config attribute would need to explicitly construct a string.
`cfg["controller"] = std::string(side_controller::human)`
instead of
`cfg["controller"] = side_controller::human`

**Comment by Vultraz** [comment]
That should probably be added to `config` if it's not there already.

**Comment by Pentarctagon** [comment]
I think that would be something for a separate PR, especially since it would need to also address simple_wml, and I have no particular desire to try touching that.

File: src/gui/dialogs/game_stats.cpp
Comment (Vultraz) [inline src/gui/dialogs/game_stats.cpp:70]: Why 5? 🤔

Comment (Pentarctagon) [reply src/gui/dialogs/game_stats.cpp:70]: I forgot to adjust it after removing the INVALID enum value.  Is there a reason to not do just `return values_.size();` though?

File: src/enum_maps/base_enum.hpp
Comment (stevecotton) [inline src/enum_maps/base_enum.hpp:35]: Please run this through clang-format (but be wary, it does bad changes to the macro definition).

**Comment by stevecotton** [comment]
Thanks, I like the direction that this is going in.

Adding a namespace and comments to the new template would be good.

**Comment by Pentarctagon** [comment]
Done.

**Comment by Vultraz** [comment]
Proposal:
```cpp
template<typename T>
struct enum_base
{
	static std::string get_string(typename T::type key)
	{
		return std::string{T::values[static_cast<int>(key)]};
	}

	static std::optional<typename T::type> get_enum(const std::string_view value)
	{
		for(unsigned int i = 0; i < T::values.size(); i++) {
			if(value == T::values[i]) {
				return static_cast<typename T::type>(i);
			}
		}

		return std::nullopt;
	}

	static constexpr std::size_t size()
	{
		return T::values.size();
	}
};

struct side_controller
{
	static constexpr std::string_view none = "null";
	static constexpr std::string_view human = "human";
	static constexpr std::string_view ai = "ai";
	static constexpr std::string_view reserved = "reserved";

	static constexpr std::array<std::string_view, 4> values{none, human, ai, reserved};
	enum class type { none = 0, human = 1, ai = 2, reserved = 3 };
};
```

This uses a data pod design instead of a derived class. `enum_base` can take any class `T` which has a `type` scoped enum and a `values` array member.

Use would be like so:
```cpp
	std::cerr << static_cast<int>(enum_base<side_controller>::get_enum("null").value()) << std::endl;
	std::cerr << enum_base<side_controller>::get_string(side_controller::type::none) << std::endl;
```

**Comment by Pentarctagon** [comment]
It's a bit more typing I suppose. Also afaik string_view is effectively incompatible with simple_wml.

**Comment by CelticMinstrel** [comment]
> Also afaik string_view is effectively incompatible with simple_wml.

simple_wml is literally based on string_view, right? Or rather, string_span, which is the same as string_view but hand-rolled. So it seems to me that making it compatible shouldn't be a lot of work.

**Comment by Pentarctagon** [comment]
Yeah, it'd be great to use string_view instead of the custom string_span, I just really don't want to need to do that as part of this PR.

**Comment by Pentarctagon** [comment]
Beyond that, I don't really have much of a preference either way, so I'll leave it up to @stevecotton .

File: src/string_enums/base_enum.hpp
Comment (stevecotton) [inline src/string_enums/base_enum.hpp:28]: Please swap the order of the words either in this line, or in the filename; otherwise we have both `enum_base` and `base_enum.hpp`.

**Comment by stevecotton** [comment]
At the moment I'm reviewing by porting src/tests/test_make_enum.cpp to use this. Would you consider adding more functions to the template class to better integrate with `config` and the iostreams operators?

> simple_wml

Having asked for more support for the normal `config` class, I'm going the opposite way for simple_wml. Assuming that anywhere that's using simple_wml is doing it for performance reasons, I think it's good to have `explicit` constructors making expensive operations look expensive.

**Comment by Pentarctagon** [comment]
> At the moment I'm reviewing by porting src/tests/test_make_enum.cpp to use this. Would you consider adding more functions to the template class to better integrate with config and the iostreams operators?

Can you be more specific?

> Having asked for more support for the normal config class, I'm going the opposite way for simple_wml. Assuming that anywhere that's using simple_wml is doing it for performance reasons, I think it's good to have explicit constructors making expensive operations look expensive.

I'm not sure what you're referring to.

**Comment by stevecotton** [comment]
> > Would you consider adding more functions to the template class to better integrate with config and the iostreams operators?
> Can you be more specific?

Adding support for
```
	std::stringstream ss;
	ss << e;
	BOOST_CHECK_EQUAL (ss.str(), "name3");
```
and
```
	config cfg;
	auto e = foo::enumname::TYPE::name;
	cfg["t1"] = e;
	cfg["t2"] = e.to_string();
```

**Comment by stevecotton** [comment]
> This uses a data pod design instead of a derived class. `enum_base` can take any class `T` which has a `type` scoped enum and a `values` array member.
> 
> Use would be like so:
> 
> ```c++
> enum_base<side_controller>::get_enum("null").value() << std::endl;
> enum_base<side_controller>::get_string(side_controller::type::none) << std::endl;
> ```

I like Vultraz's suggestion, it adds type checking to `get_string` so that the compiler catches the mistake here:
```
	first_enum::get_string(first_enum::TYPE::con3); // OK
	second_enum::get_string(first_enum::TYPE::con3); // compile error (good)
```

On top of the suggestion:
```
- class enum_base
+ struct enum_base : public T
```
and then
```
-struct side_controller : string_enums::enum_base<side_controller>
+struct side_controller_defines
 {
...
-       IMPL_GET_ENUM(TYPE)
 };
+
+using side_controller = string_enums::enum_base<side_controller_defines>;
```
That's a drop-in replacement that doesn't affect the code already in this PR that uses `side_controller`.

**Comment by Vultraz** [comment]
> On top of the suggestion:

What do we get with inheritance? 🤔 I like your idea of a type alias to avid having to write out `enum_base` every time, but I don't think we actually need it to inherit from `T`.

**Comment by stevecotton** [comment]
The inheritance means the same name, `side_controller`, in both `side_controller::get_enum(...)` and `side_controller::TYPE::ai`.

**Comment by Vultraz** [comment]
You don't need that if you have a type alias.

**Comment by Vultraz** [comment]
Yeah, I the stringstream check isn't needed since we'd be using strings/char arrays/string views directly now instead of the object.

File: src/team.hpp
Comment (Vultraz) [inline src/team.hpp:282]: This (and other similar cases) is a perfect use for `optional::value_or`, since the optional will be a nullopt if `new_controller` is invalid:
```cpp
info_.controller = side_controller::get_enum(new_controller).value_or(side_controller::TYPE::AI);
```

File: src/team.cpp
Comment (Vultraz) [inline src/team.cpp:246]: `optional::value_or` can be used here too.

Comment (Vultraz) [reply src/team.cpp:246]: Also, while this can be replaced with a one-liner, I would recommend using `auto` instead of `std::optional<side_controller::TYPE>` other places where you need a variable.

File: src/string_enums/base_enum.hpp
Comment (Pentarctagon) [reply src/string_enums/base_enum.hpp]: Done.

File: src/team.hpp
Comment (Pentarctagon) [reply src/team.hpp]: Done.

File: src/team.cpp
Comment (Pentarctagon) [reply src/team.cpp]: Done.

**Comment by Pentarctagon** [comment]
Also, for the assignment operator for config, that works currently since it's passed a struct that extends `enum_tag`, but is there a way to have it deduce the containing object's type from any particular `enum class` that's passed in? The `enum class` itself doesn't extend anything or have any implemented methods.

**Comment by Pentarctagon** [comment]
Updated the implementation to match the vultraz+octalot changes above.  A couple comments:
* I left things as `const char*`/`const std::string&` since I don't want to deal with replacing simple_wml's string_span with string_view - while on the surface it sounds simple since they are both meant to do a similar thing, everything I've ever heard about simple_wml is that it's very difficult to work with and I don't want to spend a bunch of time figuring out how it works and chasing down obscure issues unless using string_view is considered a hard requirement of merging this.
* The `public T` of `struct enum_base : public T` is there since it's what allows using, for example, `side_controller::type::NONE` rather than `side_controller_defines::type::NONE` which while not really significant in terms of extra typing is (IMO) beneficial for consistency.

So I think what remains is the two things octalot brought up in https://github.com/wesnoth/wesnoth/pull/6423#issuecomment-1017454129:
* The first wouldn't be applicable anymore since there is no longer any instance of the struct created to be provided to the `<<` operator.
* The second may (?) be possible, but to be honest I think the simplest and most elegant solution is to just call get_string where necessary rather than try to cram it into a templated method.

**Comment by CelticMinstrel** [comment]
> to be honest I think the simplest and most elegant solution is to just call get_string where necessary rather than try to cram it into a templated method.

The only possible problem with this approach is that if you _forget_ to call `get_string()`, what happens? Does it convert it to a number instead? Maybe it's fine since it's an enum class.

I definitely prefer the templated assignment operator, but if it's a compile error, I suppose it's acceptable.

**Comment by Pentarctagon** [comment]
`enum class` prevents an implicit conversion to an int, so it would be a compiler error.

File: src/string_enums/enum_base.hpp
Comment (CelticMinstrel) [inline src/string_enums/enum_base.hpp:25]: Couldn't this be enforced with a static assert? Something like…

```cpp
static_assert(T::values.size() == T::type::ENUM_MAX);
```

Although that does require adding a redundant ENUM_MAX entry to every enum, so maybe that's unacceptable…

File: src/string_enums/enum_base.hpp
Comment (CelticMinstrel) [inline src/string_enums/enum_base.hpp:50]: Come to think of it, this will break if the same string appears twice in the list… obviously there's not much to be done about that, but it could be added to the list of requirements in the doc comment at the top.

File: src/string_enums/enum_base.hpp
Comment (Pentarctagon) [reply src/string_enums/enum_base.hpp:25]: I had thought about something similar, but instead of doing an assert, define the array as:
`static constexpr std::array<const char*, static_cast<int>(type::ENUM_MAX)> values{ none, human, ai, reserved };`

I'd been trying to find it there was some built-in way in C++ to find the number of elements in an enum, but it doesn't look like such a feature exists.

Comment (CelticMinstrel) [reply src/string_enums/enum_base.hpp:25]: Yeah, the only way that I know of is adding an extra value to the enum like I mentioned.

Comment (stevecotton) [reply src/string_enums/enum_base.hpp:25]: > I'd been trying to find it there was some built-in way in C++ to find the number of elements in an enum, but it doesn't look like such a feature exists.

Looks like it may eventually be provided by the Reflection TS, but no compiler seems to support that yet.

File: src/server/wesnothd/game.hpp
Comment (stevecotton) [inline src/server/wesnothd/game.hpp:18]: Should sort later alphabetically, now that it's not called enum_maps.

Sorry for such a trivial comment, but small fixes are good.

File: src/editor/map/map_context.hpp
Comment (stevecotton) [inline src/editor/map/map_context.hpp:28]: Please put this one in its alphabetical place too.

**Comment by stevecotton** [comment]
I like it, but I also like MAKE_ENUM (at least, using it rather than trying to edit the Boost-based implementation). I'm generally in favour of merging it, but give everyone a couple more days to see if anything else comes to mind first.

> Also, barring any issues with this approach, I'd probably (over the course of 1.17) convert other uses of MAKE_ENUM to this approach with the goal of eventually removing MAKE_ENUM entirely.

I think you're committing yourself to that ;-D .

File: src/editor/map/map_context.hpp
Comment (Pentarctagon) [reply src/editor/map/map_context.hpp]: Done.

File: src/server/wesnothd/game.hpp
Comment (Pentarctagon) [reply src/server/wesnothd/game.hpp]: Done.

File: src/string_enums/enum_base.hpp
Comment (Pentarctagon) [reply src/string_enums/enum_base.hpp:51]: Done.

File: src/string_enums/enum_base.hpp
Comment (Pentarctagon) [reply src/string_enums/enum_base.hpp:25]: Added.

**Comment by Vultraz** [comment]
Posted this on Discord but adding it here too. We might consider this header lib: https://github.com/Neargye/magic_enum

**Comment by Pentarctagon** [comment]
Discussed more there, but the main issue brought up being that it requires gcc 9+, while Wesnoth's current minimum gcc version is 7+.

**Pull request merged**
